### PR TITLE
Feature #189 filter bulletins by language

### LINKED ISSUE #189: Filter index page bulletins by language
As an user i would like to filter bulletins by language

### PR DESCRIPTION
**Pull request by VilleKoivukangas**
resolves #189

### CODE REVIEW

File: edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java [hunk 1]
```diff
@@ -90,13 +101,36 @@ public int compare(DelfoiBulletin o1, DelfoiBulletin o2) {
...
+          String lang = langdetect.getLanguage(text);
+        
+          if (lang == null || !lang.equals(locale.getLanguage())) {
+            bulletinIterator.remove();
+          }
+        }
+      }
+    }
+        
+	  Long authSourceId = AuthUtils.getAuthSource("Keycloak").getId();
```
Comment (anttileppa) [inline edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java]: Tabs?

File: edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import com.optimaize.langdetect.LanguageDetectorBuilder;
+import com.optimaize.langdetect.i18n.LdLocale;
+import com.optimaize.langdetect.ngram.NgramExtractors;
+import com.optimaize.langdetect.profiles.LanguageProfile;
+import com.optimaize.langdetect.profiles.LanguageProfileReader;
+import com.optimaize.langdetect.text.CommonTextObjectFactories;
+import com.optimaize.langdetect.text.TextObject;
+import com.optimaize.langdetect.text.TextObjectFactory;
+
+public class TextLanguageDetector {
```
Comment (anttileppa) [inline edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java]: Comment this class

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
+package fi.metatavu.edelphi.languagedetect;
+
+import java.io.IOException;
+import java.util.List;
+
+import com.google.api.client.repackaged.com.google.common.base.Optional;
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java]: MINOR Remove this unused import 'com.google.api.client.repackaged.com.google.common.base.Optional'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import com.optimaize.langdetect.text.TextObjectFactory;
+
+/**
+* This class is for detecting language from a text
+*
+* @author Ville Koivukangas
+*/
+public class TextLanguageDetector {
+	
+  private static TextLanguageDetector INSTANCE = null;
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java:23]: MINOR Rename this field "INSTANCE" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3008)

File: edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  
+  public synchronized static TextLanguageDetector getInstance() throws IOException {
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/languagedetect/TextLanguageDetector.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java [hunk 6]
```diff
@@ -90,10 +101,33 @@ public int compare(DelfoiBulletin o1, DelfoiBulletin o2) {
...
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    
+      if (langdetect != null) {
+        Iterator<DelfoiBulletin> bulletinIterator = bulletins.iterator();
+        	
+        while (bulletinIterator.hasNext()) {
+          DelfoiBulletin bulletin = bulletinIterator.next();
+          String text = bulletin.getSummary().toString();
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java]: MINOR "getSummary" returns a string, there's no need to call "toString()". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1858)

File: edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java [hunk 7]
```diff
@@ -90,10 +101,33 @@ public int compare(DelfoiBulletin o1, DelfoiBulletin o2) {
         return o2.getCreated().compareTo(o1.getCreated());
       }
     });
+
+    if (!ignoreBulletinLocale) {
+      TextLanguageDetector langdetect = null;
+      try {
+        langdetect = TextLanguageDetector.getInstance();
+      } catch (IOException e) {
+        e.printStackTrace();
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java]: MINOR Use a logger to log this exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)

**Comment by sonarq** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [IndexPageController.java#L47](https://github.com/Metatavu/edelphi/blob/3507b836a96743af191480f011ed6b169ba8c564/edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java#L47): Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java [hunk 8]
```diff
@@ -90,10 +103,33 @@ public int compare(DelfoiBulletin o1, DelfoiBulletin o2) {
         return o2.getCreated().compareTo(o1.getCreated());
       }
     });
+
+    if (!ignoreBulletinLocale) {
+      TextLanguageDetector langdetect = null;
+      try {
+        langdetect = TextLanguageDetector.getInstance();
+      } catch (IOException e) {
+        System.out.println("Failed to get TextLanguageDetector instance");
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java [hunk 9]
```diff
@@ -1,13 +1,17 @@
...
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java]: MINOR Remove this unused import 'org.apache.commons.logging.Log'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java [hunk 10]
```diff
@@ -1,13 +1,17 @@
...
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log; 
+import org.apache.commons.logging.LogFactory;
```
Comment (sonarq) [inline edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java]: MINOR Remove this unused import 'org.apache.commons.logging.LogFactory'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by sonarq** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [IndexPageController.java#L49](https://github.com/Metatavu/edelphi/blob/240fe2d7568174b6851ad219d5b86932a963be19/edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java#L49): Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sonarq** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [IndexPageController.java#L50](https://github.com/Metatavu/edelphi/blob/817b2ae567e37b34a543ec1e4111f26557a12779/edelphi/src/main/java/fi/metatavu/edelphi/pages/IndexPageController.java#L50): Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**